flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/UnifyMergingChannelStateWriteRequestDispatcher.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.channel;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.runtime.state.CheckpointStorage;
+import org.apache.flink.runtime.state.CheckpointStorageAccess;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * An implementation of {@link ChannelStateWriteRequestDispatcher} which construct a build a {@link
+ * ChannelStateCheckpointWriter} that uses unified file merging mechanism to write channel state.
+ */
+class UnifyMergingChannelStateWriteRequestDispatcher
+        extends ChannelStateWriteRequestDispatcherImpl {
+
+    private final CheckpointStorageAccess checkpointStorageAccess;
+
+    UnifyMergingChannelStateWriteRequestDispatcher(
+            CheckpointStorage checkpointStorage,
+            CheckpointStorageAccess checkpointStorageAccess,
+            JobID jobID,
+            ChannelStateSerializer serializer) {
+        super(checkpointStorage, jobID, serializer);
+        this.checkpointStorageAccess = checkpointStorageAccess;
+    }
+
+    @Override
+    protected ChannelStateCheckpointWriter buildWriter(CheckpointStartRequest request)
+            throws Exception {
+        return new ChannelStateCheckpointWriter(
+                registeredSubtasks,
+                request.getCheckpointId(),
+                // checkpointStorageAccess provides unified file merging mechanism
+                checkpointStorageAccess.resolveCheckpointStorageLocation(
+                        request.getCheckpointId(), request.getLocationReference()),
```

Comment:
The only difference between `UnifyMergingChannelStateWriteRequestDispatcher` and `ChannelStateWriteRequestDispatcherImpl` is how to `resolveCheckpointStorageLocation`, right?

If so, could we not introduce `UnifyMergingChannelStateWriteRequestDispatcher`? It means we put the whole logic into `ChannelStateWriteRequestDispatcherImpl`. We can pass `unifyFileMerging` and `CheckpointStorageAccess` to `ChannelStateWriteRequestDispatcherImpl`.

The reason is I guess we only needs `unified file merging mechanism` in the future, and `ChannelStateWriteRequestDispatcherImpl` can serve `unified file merging mechanism` directly.  

WDYT?

Reply:
> The only difference between `UnifyMergingChannelStateWriteRequestDispatcher` and `ChannelStateWriteRequestDispatcherImpl` is how to `resolveCheckpointStorageLocation`, right?

Yes, it is.  

Good suggestion, I'll replace `checkpointStorage` with `checkpointStorageAccess` in  `ChannelStateWriteRequestDispatcherImpl`.